internal/monitors/replicationstatus.go

```diff
@@ -0,0 +1,61 @@
+package monitors
+
+import (
+	"log"
+	"time"
+
+	"cloudant.com/cloudant_exporter/internal/utils"
+	"github.com/IBM/cloudant-go-sdk/cloudantv1"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+)
+
+type ReplicationStatusMonitor struct {
+	Cldt     *cloudantv1.CloudantV1
+	Interval time.Duration
+	FailBox  *utils.FailBox
+}
+
+var (
+	replicatonStatus = promauto.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Name: "cloudant_replication_status",
+			Help: "Current status per replication",
+		},
+		[]string{"docid", "source", "target", "status"},
+	)
```

Comment:
The problem I see with this approach is that we'll end up with "old" statuses remaining. Like once you ahve set "running" for a replication, we're not going to remove that metric when we get to "errored" or "completed".

Not quite sure how to do this, how to set the non-current status's guage to zero. Or whether we should just drop the docid, source and target and just sum up the counts for each status as we loop. 🤔

Reply:
I don't think that is what I mean. Recall that there is a metric for every permutation of docid, source, target and status _that has been seen_. The concept of "a metric with different labels" is kinda false, what you actually have is "a metric for every combination of 'active' labels". 

I mean, if you have 2 documents. 1 running and 1 completed (that was running). 

If you sum by `status` in prometheus, it will say:

- 2 initialising replications
- 2 running replications
- 1 completed replications

Whereas what you want to see in Prometheus the "current" state of the system:

- 1 running
- 1 completed

This is because the `/metrics` output all the metrics you've created this run of the app, not just those that were "active" this cycle. Ie:

```
cloudant_replication_status{docid="aaa",source="a",target="b",status="initialising"} 1
cloudant_replication_status{docid="aaa",source="a",target="b",status="running"} 1
cloudant_replication_status{docid="aaa",source="a",target="b",status="completed"} 1
cloudant_replication_status{docid="bbb",source="a",target="b",status="initialising"} 1
cloudant_replication_status{docid="bbb",source="a",target="b",status="running"} 1
```

Whereas I believe you want:

```
cloudant_replication_status{docid="aaa",source="a",target="b",status="initialising"} 0
cloudant_replication_status{docid="aaa",source="a",target="b",status="running"} 0
cloudant_replication_status{docid="aaa",source="a",target="b",status="completed"} 1
cloudant_replication_status{docid="bbb",source="a",target="b",status="initialising"} 0
cloudant_replication_status{docid="bbb",source="a",target="b",status="running"} 1
```

That is, the `aaa` replication has the number of itself in initialising and running set to 0, because that state isn't "current" any more.

Without this, the metrics will not represent the current state, but a kind of sum of all previous states. ("You have had N replications ever in this state" rather than "you currently have these replications in this state").

In all honesty, there's also complexity in if a replication transitions state while the app isn't running, because perhaps we'd never "reset" its gauge to 0 for the old state.

Reply:
Modified to use one metric per replication status